File: src/aside/java/org/taktik/icure/logic/HealthcarePartyLogic.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.logic;

import org.taktik.icure.db.PaginatedList;
import org.taktik.icure.entities.HealthcareParty;
import org.taktik.icure.exceptions.DeletionException;
import org.taktik.icure.exceptions.DocumentNotFoundException;
import org.taktik.icure.exceptions.MissingRequirementsException;
import org.taktik.icure.db.PaginationOffset;

import javax.security.auth.login.LoginException;
import java.util.List;
import java.util.Map;

public interface HealthcarePartyLogic extends EntityPersister<HealthcareParty, String> {

	HealthcareParty getHealthcareParty(String id);


	List<HealthcareParty> findHealthcareParties(String searchString, int offset, int limit);

    Map<String,String> getHcPartyKeysForDelegate(String healthcarePartyId);

	HealthcareParty modifyHealthcareParty(HealthcareParty healthcareParty) throws MissingRequirementsException;

	List<String> deleteHealthcareParties(List<String> healthcarePartyIds) throws DeletionException;

	HealthcareParty createHealthcareParty(HealthcareParty healthcareParty) throws MissingRequirementsException;

	Map<String,String[]> updateHcPartyKeys(String healthcarePartyId, Map<String, String[]> hcPartyKeys) throws Exception;

	PaginatedList<HealthcareParty> listHealthcareParties(PaginationOffset offset, Boolean desc);

    PaginatedList<HealthcareParty> findHealthcareParties(String fuzzyName, PaginationOffset offset, Boolean desc);

	List<HealthcareParty> listByNihii(String nihii);

	List<HealthcareParty> listBySsin(String ssin);

	List<HealthcareParty> listByName(String name);

	String getPublicKey(String healthcarePartyId) throws DocumentNotFoundException;

	PaginatedList<HealthcareParty> findHealthcareParties(String type, String spec, String firstCode, String lastCode);

	List<HealthcareParty> getHealthcareParties(List<String> ids);

	List<HealthcareParty> getHealthcarePartiesByParentId(String parentId);

	PaginatedList<HealthcareParty> findHealthcarePartiesBySsinOrNihii(String searchValue, PaginationOffset<String> paginationOffset, Boolean desc);

}

File: src/main/kotlin/org/taktik/icure/asynclogic/HealthcarePartyLogic.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */

package org.taktik.icure.asynclogic

import java.net.URI
import kotlinx.coroutines.flow.Flow
import org.taktik.couchdb.DocIdentifier
import org.taktik.couchdb.ViewQueryResultEvent
import org.taktik.icure.asyncdao.HealthcarePartyDAO
import org.taktik.icure.db.PaginationOffset
import org.taktik.icure.domain.filter.chain.FilterChain
import org.taktik.icure.entities.HealthcareParty
import org.taktik.icure.entities.embed.Identifier

interface HealthcarePartyLogic : EntityPersister<HealthcareParty, String> {
	fun getGenericDAO(): HealthcarePartyDAO

	suspend fun getHealthcareParty(id: String): HealthcareParty?
	fun listHealthcarePartiesBy(searchString: String, offset: Int, limit: Int): Flow<HealthcareParty>

	@Deprecated(message = "A HCP may now have multiple AES Keys. Use getAesExchangeKeysForDelegate instead")
	suspend fun getHcPartyKeysForDelegate(healthcarePartyId: String): Map<String, String>

	suspend fun getAesExchangeKeysForDelegate(healthcarePartyId: String): Map<String, List<String>>

	suspend fun modifyHealthcareParty(healthcareParty: HealthcareParty): HealthcareParty?
	fun deleteHealthcareParties(healthcarePartyIds: List<String>): Flow<DocIdentifier>

	suspend fun createHealthcareParty(healthcareParty: HealthcareParty): HealthcareParty?

	suspend fun modifyHcPartyKeys(healthcarePartyId: String, newHcPartyKeys: Map<String, Array<String>>): Map<String, Array<String>>
	fun findHealthcarePartiesBy(offset: PaginationOffset<String>, desc: Boolean?): Flow<ViewQueryResultEvent>
	fun findHealthcarePartiesBy(fuzzyName: String, offset: PaginationOffset<String>, desc: Boolean?): Flow<ViewQueryResultEvent>
	fun listHealthcarePartiesByNihii(nihii: String): Flow<HealthcareParty>
	fun listHealthcarePartiesBySsin(ssin: String): Flow<HealthcareParty>
	fun listHealthcarePartiesByName(name: String): Flow<HealthcareParty>

	suspend fun getPublicKey(healthcarePartyId: String): String?
	fun listHealthcarePartiesBy(type: String, spec: String, firstCode: String, lastCode: String): Flow<ViewQueryResultEvent>
	fun getHealthcareParties(ids: List<String>): Flow<HealthcareParty>
	fun findHealthcarePartiesBySsinOrNihii(searchValue: String, paginationOffset: PaginationOffset<String>, desc: Boolean): Flow<ViewQueryResultEvent>
	fun getHealthcarePartiesByParentId(parentId: String): Flow<HealthcareParty>
	suspend fun getHcpHierarchyIds(sender: HealthcareParty): HashSet<String>
	suspend fun createHealthcarePartyOnUserDb(healthcareParty: HealthcareParty, HealthcareParty: URI): HealthcareParty?
	fun filterHealthcareParties(paginationOffset: PaginationOffset<Nothing>, filter: FilterChain<HealthcareParty>): Flow<ViewQueryResultEvent>
	fun listHealthcarePartyIdsByIdentifiers(hcpIdentifiers: List<Identifier>): Flow<String>
	fun listHealthcarePartyIdsByCode(codeType: String?, codeCode: String?, startValueDate: Long?, endValueDate: Long?): Flow<String>
	fun listHealthcarePartyIdsByTag(tagType: String?, tagCode: String?, startValueDate: Long?, endValueDate: Long?): Flow<String>
}
